source/common/router/config_impl.cc

```diff
@@ -1003,7 +1003,8 @@ RouteConstSharedPtr VirtualHostImpl::getRouteFromEntries(const Http::HeaderMap&
   if (ssl_requirements_ == SslRequirements::ALL && forwarded_proto_header->value() != "https") {
     return SSL_REDIRECT_ROUTE;
   } else if (ssl_requirements_ == SslRequirements::EXTERNAL_ONLY &&
-             forwarded_proto_header->value() != "https" && !headers.EnvoyInternalRequest()) {
+             forwarded_proto_header->value() != "https" &&
+             not Http::HeaderUtility::isEnvoyInternalRequest(headers)) {
```

Comment:
This is the core fix

Reply:
LGTM. Nit: please use standard predicate syntax `!` instead of this new-fangled one that is pretty confusing to most C/C++ folks.